File: Leetcode-one-by-one/src/Concatenation of Consecutive Binary Numbers.swift
class Solution {
    func concatenatedBinary(_ n: Int) -> Int {
        let MOD = 1_000_000_007
        var result = 0
        var bitLength = 0
        
        for i in 1...n {
            // If i is power of 2, increase bit length
            if (i & (i - 1)) == 0 {
                bitLength += 1
            }
            
            // Shift left and add current number
            result = ((result << bitLength) % MOD + i) % MOD
        }
        
        return result
    }
}

File: Leetcode-one-by-one/src/Concatenation of Consecutive Binary Numbers.java
class Solution {
    public int concatenatedBinary(int n) {
        final int MOD = 1_000_000_007;
        long result = 0;
        int bitLength = 0;

        for (int i = 1; i <= n; i++) {
            // If i is a power of 2, increment bit length
            if ((i & (i - 1)) == 0) {
                bitLength++;
            }

            // Shift left by bitLength and add i
            result = ((result << bitLength) + i) % MOD;
        }

        return (int) result;
    }
}
